File: leetcode/python/735_asteroid_collision.py
from inspect import stack


class Solution:
    def asteroidCollision(self, asteroids: list[int]) -> list[int]:
        stack: list[int] = []
        for n in asteroids:
            while n < 0 and stack and stack[-1] > 0:
                if stack[-1] < -n:
                    stack.pop()
                    continue
                elif stack[-1] == -n:
                    stack.pop()
                break
            else:
                stack.append(n)
        return stack

s = Solution()
print(s.asteroidCollision([5,10,-5]))
print(s.asteroidCollision([8,-8]))
print(s.asteroidCollision([10, 2,-5]))

File: leetcode/rust/src/735_asteroid_collision.rs
pub fn asteroid_collision(asteroids: Vec<i32>) -> Vec<i32> {
    let mut stack = vec![];
    for n in asteroids {
        if n > 0 {
            stack.push(n);
            continue;
        }

        let mut exploded = false;
        while !stack.is_empty() && stack.last().unwrap() > &0 {
            let last = stack.pop().unwrap();
            if last >= -n {
                if last > -n {
                    stack.push(last);
                }
                exploded = true;
                break;
            }
        }

        if !exploded {
            stack.push(n);
        }
    }
    return stack;
}
